content/advent-2017/property-based-testing.md

```diff
@@ -0,0 +1,340 @@
+
+After a while, I started noticing that the class of edge cases I thought he'd fixed were coming back. That's when I went back to the code only to discover that the tests have passed because the algorithm was fixed to only include the specific edge case I had specified, and nothing else.
+
+Along the way I discovered other things that made me really quite upset (mostly at myself for not reviewing the code carefully). Here's what happened. Imagine I asked you to test the `Add` function from above. What is one valid way to make sure it always passes? Here's one way:
+
+```go
+func TestAdd(t *testing.T) {
+	a := 1
+	b := 2
+	out := Add(a, b)
+	if out != a + b {
+		t.Errorf("Add failed...")
+	}
+}
+```
+
+If you can't tell what's wrong, here's a less subtle version of it:
+
+```go
+func TestAdd(t *testing.T) {
+	a := 1
+	b := 2
+	out := Add(a, b)
+	if out != Add(a, b) {
+		t.Errorf("Add failed...")
+	}
+}
+```
+
+If you're testing function `Foo`, you can't use `Foo` to verify that `Foo` works! That's just nuts!
+
+# Testing On More Inputs #
+
+"Be robust to various inputs" means that you have to test for a lot of different inputs. OK codebases will have tests for the happy path, where everything  goes well. Better code bases will have test cases where the program fails - the so-called tragic path - to test the ability of the program to handle failure conditions correctly. Being able to correctly handle failure conditions is the prerequisite to robust software.
+
+Testing along the tragic path is not easy. For many functions and programs it's difficult to reason out where it may go wrong. In Go, this is somewhat alleviated by making errors a value. Even so, a codebase may often choose to ignore error checking:
+
+```go
+func returnsError() error { return errors.New("This Error") }
+
+func foo() {
+	returnsError() // bad! Don't do this! Always check for and handle errors!
+}
+```
+
+But let's say you are a diligent programmer, and you checked all the errors, other forms of errors may occur too: panics from slice indices being out of bounds, etc. 
+
+So an idea emerges: why don't we test on all possible values? Or at least as many values as we can? 
+
+Enter the notion of [fuzz testing](https://en.wikipedia.org/wiki/Fuzzing): we feed the program random inputs, and then watch for it to fail. Fuzz testing often leads one to find subtle bugs you don't expect. Here's one that Chris Marshall discovered on my [skiprope](https://github.com/chewxy/skiprope) library. I have to this day, no idea what the correct fix is. Solutions welcome
```

Comment:
Missing link to actual `skiprope` issue?

Reply:
thanks for the spot. Will be in next commit